From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Christian Bourjau <christian.bourjau@quantco.com>
Date: Tue, 13 Oct 2026 22:58:37 +0200
Subject: [PATCH] Use a preinstalled libcoremltools for the CoreML EP

libcoremltools provides the generated CoreML protobuf classes and the
MILBlob/modelpackage helpers. Link it instead of fetching the coremltools
sources and compiling a second copy of them.
---
 .../external/onnxruntime_external_deps.cmake  | 14 ++--
 cmake/onnxruntime_providers_coreml.cmake      | 77 ++-----------------
 2 files changed, 13 insertions(+), 78 deletions(-)

diff --git a/cmake/external/onnxruntime_external_deps.cmake b/cmake/external/onnxruntime_external_deps.cmake
index 1208e8a..73276f0 100644
--- a/cmake/external/onnxruntime_external_deps.cmake
+++ b/cmake/external/onnxruntime_external_deps.cmake
@@ -1035,15 +1035,11 @@ if(onnxruntime_USE_COREML)
   set(FP16_BUILD_BENCHMARKS OFF CACHE INTERNAL "")
   onnxruntime_fetchcontent_makeavailable(fp16)
 
-  onnxruntime_fetchcontent_declare(
-    coremltools
-    URL ${DEP_URL_coremltools}
-    URL_HASH SHA1=${DEP_SHA1_coremltools}
-    PATCH_COMMAND ${Patch_EXECUTABLE} --binary --ignore-whitespace -p1 < ${PROJECT_SOURCE_DIR}/patches/coremltools/crossplatformbuild.patch
-    EXCLUDE_FROM_ALL
-  )
-  # we don't build directly so use Populate. selected files are built from onnxruntime_providers_coreml.cmake
-  FetchContent_Populate(coremltools)
+  # Use a preinstalled libcoremltools. It provides the generated CoreML protobuf classes
+  # and the MILBlob/modelpackage helpers, along with their headers.
+  find_library(coremltools_LIBRARY NAMES coremltools REQUIRED)
+  find_path(coremltools_SOURCE_DIR NAMES mlmodel/build/format/Model.pb.h PATH_SUFFIXES coremltools REQUIRED)
+  message(STATUS "Using coremltools library ${coremltools_LIBRARY} with headers in ${coremltools_SOURCE_DIR}")
 
 endif()
 
diff --git a/cmake/onnxruntime_providers_coreml.cmake b/cmake/onnxruntime_providers_coreml.cmake
index f3425b9..9df86a9 100644
--- a/cmake/onnxruntime_providers_coreml.cmake
+++ b/cmake/onnxruntime_providers_coreml.cmake
@@ -19,47 +19,14 @@ if(LINUX)
   endif()
 endif()
 
-# Compile CoreML proto definition to ${CMAKE_CURRENT_BINARY_DIR}/coreml_proto
-set(COREML_PROTO_ROOT ${coremltools_SOURCE_DIR}/mlmodel/format)
-file(GLOB coreml_proto_srcs "${COREML_PROTO_ROOT}/*.proto")
-
-onnxruntime_add_static_library(coreml_proto ${coreml_proto_srcs})
-target_include_directories(coreml_proto
-                           PUBLIC $<TARGET_PROPERTY:${PROTOBUF_LIB},INTERFACE_INCLUDE_DIRECTORIES>
-                           $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>)
-target_compile_definitions(coreml_proto
-                           PUBLIC $<TARGET_PROPERTY:${PROTOBUF_LIB},INTERFACE_COMPILE_DEFINITIONS>)
-set_target_properties(coreml_proto PROPERTIES COMPILE_FLAGS "-fvisibility=hidden")
-set_target_properties(coreml_proto PROPERTIES COMPILE_FLAGS "-fvisibility-inlines-hidden")
-
-set(_src_sub_dir "coreml_proto/")
-onnxruntime_protobuf_generate(
-  APPEND_PATH
-  GEN_SRC_SUB_DIR ${_src_sub_dir}
-  IMPORT_DIRS ${COREML_PROTO_ROOT}
-  TARGET coreml_proto
-)
-
-if (NOT onnxruntime_BUILD_SHARED_LIB)
-  install(TARGETS coreml_proto
-          EXPORT ${PROJECT_NAME}Targets
-          ARCHIVE   DESTINATION ${CMAKE_INSTALL_LIBDIR}
-          LIBRARY   DESTINATION ${CMAKE_INSTALL_LIBDIR}
-          RUNTIME   DESTINATION ${CMAKE_INSTALL_BINDIR}
-          FRAMEWORK DESTINATION ${CMAKE_INSTALL_BINDIR}
-  )
-endif()
-
-# Add the .proto and generated .cc/.h files to the External/coreml_proto folder in Visual Studio.
-# Separate source_group for each as the .proto files are in the repo and the .cc/.h files are generated in the build
-# output directory.
-set_target_properties(coreml_proto PROPERTIES FOLDER "External")
-source_group(TREE ${COREML_PROTO_ROOT} PREFIX coreml_proto FILES ${coreml_proto_srcs})
-
-# filter to the generated .cc/.h files
-get_target_property(coreml_proto_generated_srcs coreml_proto SOURCES)
-list(FILTER coreml_proto_generated_srcs INCLUDE REGEX "\.pb\.(h|cc)$")
-source_group(TREE ${CMAKE_CURRENT_BINARY_DIR} PREFIX coreml_proto_generated FILES ${coreml_proto_generated_srcs})
+# libcoremltools provides the generated CoreML protobuf classes. ONNX Runtime includes
+# them as coreml_proto/*.pb.h, so expose their directory under that name.
+set(_coreml_proto_include_dir ${CMAKE_CURRENT_BINARY_DIR}/coreml_proto_include)
+file(MAKE_DIRECTORY ${_coreml_proto_include_dir})
+file(CREATE_LINK ${coremltools_SOURCE_DIR}/mlmodel/build/format ${_coreml_proto_include_dir}/coreml_proto SYMBOLIC)
+add_library(coreml_proto INTERFACE)
+target_include_directories(coreml_proto INTERFACE ${_coreml_proto_include_dir})
+target_link_libraries(coreml_proto INTERFACE ${coremltools_LIBRARY} ${PROTOBUF_LIB})
 
 # These are shared utils,
 # TODO, move this to a separate lib when used by EPs other than NNAPI and CoreML
@@ -85,33 +52,6 @@ file(GLOB_RECURSE
   "${ONNXRUNTIME_ROOT}/core/providers/coreml/builders/*.cc"
 )
 
-  # Add helpers to create mlpackage weights. limit to just the files we need to minimize the changes to make them
-  # build on Windows and Linux.
-file(GLOB
-    onnxruntime_providers_coreml_milblob_cc_srcs CONFIGURE_DEPENDS
-    "${coremltools_SOURCE_DIR}/mlmodel/src/MILBlob/*.hpp"
-    "${coremltools_SOURCE_DIR}/mlmodel/src/MILBlob/*.cpp"
-    "${coremltools_SOURCE_DIR}/mlmodel/src/MILBlob/Util/*.hpp"
-    "${coremltools_SOURCE_DIR}/mlmodel/src/MILBlob/Blob/BlobDataType.hpp"
-    "${coremltools_SOURCE_DIR}/mlmodel/src/MILBlob/Blob/StorageFormat.hpp"
-    "${coremltools_SOURCE_DIR}/mlmodel/src/MILBlob/Blob/FileWriter.?pp"
-    "${coremltools_SOURCE_DIR}/mlmodel/src/MILBlob/Blob/StorageWriter.?pp"
-)
-
-# Add helpers to create mlpackage
-file(GLOB
-    onnxruntime_providers_coreml_modelpackage_cc_srcs CONFIGURE_DEPENDS
-    "${coremltools_SOURCE_DIR}/modelpackage/src/ModelPackage.?pp"
-    "${coremltools_SOURCE_DIR}/modelpackage/src/utils/JsonMap.?pp"
-)
-
-set(coremltools_srcs
-    ${onnxruntime_providers_coreml_milblob_cc_srcs}
-    ${onnxruntime_providers_coreml_modelpackage_cc_srcs}
-)
-
-source_group(TREE ${coremltools_SOURCE_DIR} PREFIX coremltools FILES ${coremltools_srcs})
-
 # Add CoreML objective c++ source code
 if (APPLE)
   file(GLOB
@@ -148,7 +88,6 @@ source_group(TREE ${ONNXRUNTIME_INCLUDE_DIR} FILES ${onnxruntime_providers_corem
 onnxruntime_add_static_library(onnxruntime_providers_coreml
   ${onnxruntime_providers_coreml_public_headers}
   ${onnxruntime_providers_coreml_cc_srcs}
-  ${coremltools_srcs}
 )
 
 onnxruntime_add_include_to_target(onnxruntime_providers_coreml
